File: src/personal_data_warehouse/objectstore/factory.py
"""Backend selection for the object storage abstraction.

Application call sites build an :class:`ObjectStoreSpec` and call
:func:`build_object_store`; they never import a concrete backend. Adding a
backend means registering one builder in ``_BUILDERS`` -- no call site changes.
"""

from __future__ import annotations

from collections.abc import Callable

from personal_data_warehouse.objectstore.base import ObjectStore, ObjectStoreError
from personal_data_warehouse.objectstore.spec import (
    GOOGLE_DRIVE_BACKEND,
    S3_BACKEND,
    SUPPORTED_BACKENDS,
    ObjectStoreSpec,
)


def build_object_store(spec: ObjectStoreSpec, *, settings=None, service=None) -> ObjectStore:
    """Construct the object store described by ``spec``.

    ``settings`` is required for backends that build their own client from
    credentials (Google Drive). ``service`` lets callers (and tests) inject a
    pre-built backend client instead of constructing one.
    """

    builder = _BUILDERS.get(spec.backend)
    if builder is None:
        supported = ", ".join(SUPPORTED_BACKENDS)
        raise ObjectStoreError(
            f"Unsupported object storage backend: {spec.backend!r} (supported: {supported})"
        )
    return builder(spec, settings, service)


def _build_google_drive(spec: ObjectStoreSpec, settings, service) -> ObjectStore:
    # Imported lazily so that importing the factory does not require the Google
    # client libraries unless the Drive backend is actually used.
    from personal_data_warehouse.objectstore.google_drive import (
        GoogleDriveObjectStore,
        build_google_drive_service,
    )

    if not spec.namespace:
        raise ObjectStoreError("Google Drive object store requires a folder id (spec.namespace)")
    if service is None:
        connection = spec.google_drive
        if connection is None:
            raise ObjectStoreError("Google Drive object store requires spec.google_drive connection")
        if settings is None:
            raise ObjectStoreError("Google Drive object store requires settings to build a service")
        service = build_google_drive_service(
            account=connection.account,
            settings=settings,
            request_timeout_seconds=connection.request_timeout_seconds,
        )
    return GoogleDriveObjectStore(
        folder_id=spec.namespace,
        service=service,
        source=spec.source,
        legacy_sources=spec.legacy_sources,
        audio_kind=spec.blob_kind,
        metadata_kind=spec.metadata_kind,
    )


def _build_s3(spec: ObjectStoreSpec, settings, service) -> ObjectStore:
    from personal_data_warehouse.objectstore.s3 import S3ObjectStore

    if service is not None:
        return service
    if spec.s3 is None:
        raise ObjectStoreError("S3 object store requires spec.s3 connection")
    return S3ObjectStore(connection=spec.s3)


_BUILDERS: dict[str, Callable[[ObjectStoreSpec, object, object], ObjectStore]] = {
    GOOGLE_DRIVE_BACKEND: _build_google_drive,
    S3_BACKEND: _build_s3,
}

File: app/internal/objectstore/factory.go
package objectstore

import (
	"errors"
	"fmt"
)

// Backend names. Mirrors the Python SUPPORTED_BACKENDS.
const (
	GoogleDriveBackend = "google_drive"
	S3Backend          = "s3"
)

// SupportedBackends lists the backends the factory knows about.
var SupportedBackends = []string{GoogleDriveBackend, S3Backend}

// ErrS3BucketRequired is returned when an S3 spec omits the bucket.
var ErrS3BucketRequired = errors.New("objectstore: S3Connection.Bucket is required")

// GoogleDriveConnection carries the Drive connection for a Spec. HTTPClient must
// be authenticated for the Drive scope (see HTTPClientFromAuthorizedUserJSON);
// it is provided by the caller so this package never builds Drive credentials
// implicitly. APIBaseURL/UploadBaseURL default to Google and exist for tests.
type GoogleDriveConnection struct {
	HTTPClient    httpDoer
	APIBaseURL    string
	UploadBaseURL string
}

// Spec is the backend-neutral construction description handed to
// BuildObjectStore. It mirrors the Python ObjectStoreSpec.
type Spec struct {
	Backend       string
	Namespace     string // Drive folder id for managed stores, or S3 bucket/prefix root
	Source        string
	BlobKind      string
	MetadataKind  string
	LegacySources []string
	GoogleDrive   *GoogleDriveConnection
	S3            *S3Connection
}

// GoogleDriveSpec builds a Drive Spec. The httpClient must be authenticated for
// the Drive scope.
func GoogleDriveSpec(folderID, source, blobKind, metadataKind string, legacySources []string, conn GoogleDriveConnection) Spec {
	return Spec{
		Backend:       GoogleDriveBackend,
		Namespace:     folderID,
		Source:        source,
		BlobKind:      blobKind,
		MetadataKind:  metadataKind,
		LegacySources: legacySources,
		GoogleDrive:   &conn,
	}
}

// BuildObjectStore constructs the ObjectStore described by spec. Adding a
// backend means adding a case here; no call site changes.
func BuildObjectStore(spec Spec) (ObjectStore, error) {
	switch spec.Backend {
	case GoogleDriveBackend:
		if spec.GoogleDrive == nil {
			return nil, fmt.Errorf("objectstore: google_drive backend requires spec.GoogleDrive connection")
		}
		return NewGoogleDriveStore(GoogleDriveOptions{
			FolderID:      spec.Namespace,
			HTTPClient:    spec.GoogleDrive.HTTPClient,
			Source:        spec.Source,
			LegacySources: spec.LegacySources,
			AudioKind:     spec.BlobKind,
			MetadataKind:  spec.MetadataKind,
			APIBaseURL:    spec.GoogleDrive.APIBaseURL,
			UploadBaseURL: spec.GoogleDrive.UploadBaseURL,
		}), nil
	case S3Backend:
		if spec.S3 == nil {
			return nil, fmt.Errorf("objectstore: s3 backend requires spec.S3 connection")
		}
		return NewS3Store(*spec.S3)
	default:
		return nil, fmt.Errorf("objectstore: unsupported backend %q (supported: %v)", spec.Backend, SupportedBackends)
	}
}
